src/engine.rs

```diff
@@ -842,19 +846,41 @@ impl<'a, 'b> Cpu<'a, 'b> {
                         self.hartid, self.state.pc, addr
                     );
                 }
-                // trace!("Load 0x{:x} ({}B)", addr, 8 << size);
-                self.engine
+                let word_offset = addr % 4;
+                let mask = (!(u64::MAX << (8 << size))) as u32;
+                if (size as u32) > (4 - word_offset) {
```

Comment:
This will not always work, will it? Allowed are only:
- size = 0 --> Any addresses
- size = 1 --> word_offset = (0,2)
- size = 2 --> word_offset = 0

So a word_offset of 3 with size 1 or word_offset of 1,2 with size 2 will not trigger the warning. But checking whether the size LSBs are zero should work. So `word_offset % (1<<size) == 0` could work.

Also, unless I am mistaken, we should never have unaligned loads anywhere, so can you directly move the check to run for all memory locations and also for the store function?

Reply:
It definitely makes sense to assert that the size/offset combination is valid but I'm a bit confused with your last sentence. Unaligned load and store are possible (in the DMA for instance) as long as they are not across two words. Are we aligned on this or am I missing something?

Reply:
I don't think that the proposed approach for the assertion works. For instance, if `word_offset` is 1 and `size` is 1 we have `1 % 2`.

I think that when `(4 - word_offset) < (1<<size)` is `True` then the load/store is invalid.

Reply:
Addressed in [this commit](https://github.com/pulp-platform/banshee/pull/18/commits/95619028583909da22c571e6fdbf3cd400d35545).